[#7690] additional details about frontend interactions

smotornyuk opened the pull request:
* Fix fronted installation docs and explain why `npm ci` is better than `npm i`
* Mention gulpfile.js in the docs and add a link to gulp docs
* update gulp tasks, including datastore's highlight.js into scope

wardi commented:
highlight.js for the datastore needs some languages not in the default install (powershell and R) so I had built a custom one with only the languages required. How is that done now?

smotornyuk commented:
I explained the process in #7690. 
As for highlight.js - I'm not sure if we are doing it in the right way. First of all: a short list of supported languages. Now we use highlight.js with support of `javascript`, `python`, `R`, `PowerShell`, and cURL. If somebody extends the data API snippet and adds Ruby, this snippet won't be highlighted. The person will be forced to compile another version of highlight.js and include it in addition to our default version. 
The second problem: the text_view plugin. It also uses highlight.js. This PR will copy the same highlight.js source into the datastore and text view. But if both plugins are enabled, we'll have two different assets with the same content. I don't think that the data API button can be rendered for the text view, so we won't load them simultaneously. But still, it seems that defining the highlight.js asset inside CKAN core and using it in datastore and text view as a dependency is better than registering it twice.

In the original issue I explained how we can build custom version of highlight.js using 2 `npm run ...` aliases. How about creating a default `ckan/highlight.js` asset with support of languages that are used out of the box by CKAN snippets(datastore and textview) and explaining the way to build a different version of `highlight.js` and overriding `ckan/highlight.js` asset in extension, so that there will be always only one version of highlight.js loaded into the app and users are still allowed to add new languages

pdelboca commented:
I love having a Makefile, but can we remove `gulp` as well? If not we keep adding layers of automation over layers of automation for our build process.

Can we rely on the Makefile for our CSS build as well?

smotornyuk commented:
We can, but gulp does it much better, than bash. If we want to compile SCSS into CSS, bash will work. But gulp shines when we are using watchers or including sourcemaps into the development build of static files. 
So I think that keeping gulp worth the effort. But it should be used only for CSS/JS transformation and never for maintenance tasks that are not modifying frontend files.

I'll follow up with the second comment summarizing changes and ideas behind them

smotornyuk commented:
The scope was changed after discussion and that's what we have now.

Using `npm` for updating vendor libraries is not very clear for CKAN developers, because they are usually more familiar with python and bash than nodejs and its package manager. And using Gulp for copying over files doesn't sound like a proper solution anyway, so let's try a different approach.

In this PR I included a `Makefile`:
![image](https://github.com/ckan/ckan/assets/11091199/9add6eb8-0750-4343-96a0-471831ac7096)

It contains a rule that copies vendor libraries from `node_modules` into public CKAN folders: `make vendor-copy`. 

But why use just a single rule if we have `Makefile`? I checked our release guide and added a number of additional rules for building the documentation, pulling/pushing translations, etc.

Eventually, we can encapsulate most of the non-click scripts inside a `Makefile`. It's more familiar than Gulp and suits better for generic tasks that do not include SCSS compilation. Maybe even the whole release process can be implemented as an interactive `make release` rule(but I don't want to go this far in the current PR).


@wardi, as for the original problem, multiple versions of highligh.js are included in the CKAN source, it is also handled via make-rule.
`make vendor-copy`, which copies updated vendor libraries into a public folder, automatically calls `hljs-build` rule. This rule installs all the build dependencies for Highlight.js(yes, they have not been installed automatically because of hljs policy) and builds Higlight.js for the predefined set of languages. If we want to extend this set, it can be done via one change inside Makefile.
Instead of copying Highlight.js into text_view and datastore plugin, I created `vendor/hljs` asset that is used by both plugins and can be potentially used by any external extension. In this way, we always have only one version of Highlight.js and don't see any errors in the JS console.
I updated documentation, explaining the way to include a new language support or a new theme for highlight.js, so every problem is covered.


As @pdelboca suggested, we can even go further and remove Gulp completely. But implementing a file watcher that minifies/includes sourcemaps into build CSS files is not a trivial task for bash and I afraid we'll end up with unreadable make-rule. That's why I suggest keeping Gulp, even though it adds extra layer to our toolset.
Or we can remove gulp and implement just a single make-rule that directly compiles SCSS into CSS. It will work in most cases, as usually we are not changing styles much. But the first person who is going to update/debug/rewrite CKAN theme will curse us a lot:)

pdelboca commented:
@smotornyuk great work! We can pilot this for the 2.11 release.

I'm still eager to remove gulp completely and let the future person in charge of UX start fresh, but I won't insist :D

smotornyuk commented:
Apart from the karma damage, personally, I have no reason to insist on any option.

Below are commands that can be used instead of gulpfile (`npm run build` and `npm run watch`). I mentioned at the meeting, that currently we are not using any fancy features of gulp, like media-query merging/sorting, so it's possible to replicate the current behavior. 

```Makefile

frontent-build: ## compile frontend
  npx sass --no-source-map \
    ckan/public/base/scss/main.scss:ckan/public/base/css/main.css \
    ckan/public/base/scss/main-rtl.scss:ckan/public/base/css/main-rtl.css 

frontent-watch:  ## Watch stylesheets and recompile when they change(with sourcemaps)
  npx sass \
    ckan/public/base/scss/main.scss:ckan/public/base/css/main.css \
    ckan/public/base/scss/main-rtl.scss:ckan/public/base/css/main-rtl.css 
```

I can include them in PR, if nobody is using Gulp and we agree to give the person who will work on frontend redevelopment a freedom to choose a toolset

wardi commented on doc/contributing/frontend/index.rst:
```diff
@@ -373,3 +392,100 @@ available in all test contexts. Tests can be made asynchronous by using promises
+  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/styles/docco.min.css">
+
+
+To reduce its size, Highlight.js was compileled for small subset of supported
```
```suggestion
To reduce its size, Highlight.js was compiled for small subset of supported
```

wardi reviewed (approved):
approving but I'll defer to @pdelboca on gulp usage, I have no strong opinion either way

wardi commented:
my updated opinion on gulp after having used it is definitely "not a fan" but it's better than nothing. @smotornyuk would you resolve the conflicts so we can merge?
